Some data task signposts are not ended properly

bnham opened the pull request:
#### 6fb617ebe42bf0dfa70e4936776152d667990351
<pre>
Some data task signposts are not ended properly
<a href="https://bugs.webkit.org/show_bug.cgi?id=251630">https://bugs.webkit.org/show_bug.cgi?id=251630</a>
rdar://104976161

Reviewed by Alex Christensen.

Sometimes, when a data task is cancelled, the didCompleteWithError method is not called. Since that
is where we end data task signposts, this can lead to dropped signposts and incorrectly attributed
resource loads in perf traces.

To fix this, move the end signpost call to the destructor instead. This matches the begin signpost
call, which occurs in the constructor.

Note that this only affects behavior when a perf tracing tool is explicitly enabled; these signposts
are disabled otherwise.

* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::~NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::didCompleteWithError):

Canonical link: <a href="https://commits.webkit.org/259826@main">https://commits.webkit.org/259826@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36b97c9e2a3d3c49cfbecc02ea6059a43a8d7e04

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105930 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14996 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38774 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115120 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16415 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6184 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98154 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114872 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111680 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12495 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95469 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40038 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94360 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27113 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81703 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8257 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28466 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8742 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5041 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14370 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48008 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6798 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10293 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 reviewed (approved):
This is fine, but hopefully there's not much time between didCompleteWithError and the destructor in the normal case.

bnham committed:
Some data task signposts are not ended properly
https://bugs.webkit.org/show_bug.cgi?id=251630
rdar://104976161

Reviewed by Alex Christensen.

Sometimes, when a data task is cancelled, the didCompleteWithError method is not called. Since that
is where we end data task signposts, this can lead to dropped signposts and incorrectly attributed
resource loads in perf traces.

To fix this, move the end signpost call to the destructor instead. This matches the begin signpost
call, which occurs in the constructor.

Note that this only affects behavior when a perf tracing tool is explicitly enabled; these signposts
are disabled otherwise.

* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::~NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::didCompleteWithError):

Canonical link: https://commits.webkit.org/259826@main
